test/unit/math/mix/mat/fun/cov_periodic_test.cpp

```diff
@@ -0,0 +1,10872 @@
+    x_rvec_d[i].resize(1,5);
+    x_rvec_d[i] << 1, 2, 3, 4, 5;
+    x_rvec_fv[i].resize(1,5);
+    x_rvec_fv[i] << 1, 2, 3, 4, 5;
+  }
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma_d, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma, d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma_d, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma_d, l, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma, l_d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_d, sigma_d, l_d, p_d));
+
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma_d, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma, d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma_d, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma_d, l, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma, l_d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_d, sigma_d, l_d, p_d));
+
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma_d, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma, d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma_d, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma_d, l, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma, l_d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_vec_fv, sigma_d, l_d, p_d));
+
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma_d, l, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma, d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma_d, l_d, p));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma_d, l, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma, l_d, p_d));
+  EXPECT_NO_THROW(cov_periodic(x_rvec_fv, sigma_d, l_d, p_d));
+}
+
+TEST(AgradMixMath, ffv_cov_periodic1) {
+  using stan::math::cov_periodic;
+  using stan::math::fvar;
+  using stan::math::var;
+
+  fvar<fvar<var> > a(1.0,1.0);
+  fvar<fvar<var> > b(2.0,1.0);
+  fvar<fvar<var> > c(3.0,1.0);
+  fvar<fvar<var> > d(4.0,1.0);
+  fvar<fvar<var> > e(0.0,1.0);
+
+  std::vector<fvar<fvar<var> > > x;
+  x.push_back(e);
+  x.push_back(a);
+  x.push_back(b);
+  x.push_back(c);
+  x.push_back(d);
+
+  fvar<fvar<var> > sigma(2.0, 1.0);
+  fvar<fvar<var> > l(1.0, 1.0);
+  fvar<fvar<var> > p(0.7, 1.0);
+  double sigma_d(2.0);
+  double l_d(1.0);
+  double p_d(0.7);
+
+  std::vector<double> x_d;
+  x_d.push_back(1);
+  x_d.push_back(2);
+  x_d.push_back(3);
+  x_d.push_back(4);
+  x_d.push_back(5);
+
+  Eigen::Matrix<fvar<fvar<var> >, -1, -1> cov;
+  EXPECT_NO_THROW(cov = cov_periodic(x, sigma, l, p));
+  EXPECT_NO_THROW(cov_periodic(x, sigma_d, l, p));
```

Comment:
I had a look through the tests in this file. Most of it seems to be initializing and trying arguments of different combinations of types -- not really testing the specifics of the periodic covariance.

Is there any way to make these tests generic for other GPs that someone might drag along?

I was working on vectorizing the _rng functions with Sean/Bob. In that case too there's a ton of generic boilerplate. What we ended up doing on that front was making little test rigs (https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/prob/normal_test.cpp for instance) that wrap around the _rng we're trying to test and then feed that through some big fancy templated machinery that instantiates all these annoying type tests.

We cleaned up a lot of the super verbose lists of combinations of types with some C++ typelist template magic as well. Here's the before: https://github.com/stan-dev/math/blob/72eb95060662885a82a218d0c296c3c9e87b6286/test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp#L270 . Here's the after: https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp#L283

Do you think something like this is possible here?

My apologies for talking about working on more tests haha. I imagine you're quite sick of them after writing this much stuff :D.